src/Ampersand/Input/PreProcessor.hs

```diff
@@ -0,0 +1,110 @@
+module Ampersand.Input.PreProcessor (
+      preProcess
+    , PreProcDefine
+) where
+
+import Data.List
+import Data.String
+import Data.Char
+import Data.Either
+import Data.Maybe
+import Prelude
+
+type PreProcDefine = String
+
+preProcess :: [PreProcDefine] -> String -> String
+preProcess defs = block2file defs True . file2block
+ 
+-- Run the full parser
+file2block :: String -> Block
+file2block = parseBlock . map parseLine . lines
+
+-- Turn a block back
+block2file :: [PreProcDefine] -> Bool -> Block -> String
+block2file defs shown = unlines . map (blockElem2string defs shown)
+
+-- Handle single entry in a block
+blockElem2string :: [PreProcDefine] -> Bool -> BlockElem -> String
+blockElem2string _    True  (Left  line) = line
+blockElem2string _    False (Left  line) = "-- hide by preprocc " ++ line
+-- Lots of unpacking to get to the IfBlock
+blockElem2string defs False (Right (IfBlock (Guard guard) block)) = 
+    "-- IF " ++ guard ++ block2file defs False block ++ "-- ENDIF"
+blockElem2string defs True  (Right (IfBlock (Guard guard) block)) = 
+    "-- IF " ++ guard ++ "\n" ++ 
+    (block2file defs (guard `elem` defs) block) ++ 
+    "\n-- ENDIF"
+ 
+-- Here the experimentation starts
+
+{- Our grammar is:
+Codeline = Any line that is not IfStartLine or IfEndLine
+Word = consecutive non-whitespace chars
+String litterals are encased in pairs of '
+: is concatenation
+| is disjunction
+\ is negation
+
+IfStartLine = ' '* : '-- if ' : Word : ' '* : '\n'
```

Comment:
This part should best go into the Lexer. Define three extra constructors at LexterToken.hs:

```
Lexeme = ...

   | PreCommonLine String -- ^ A common line for the preprocessor
   | PreIfStart   String -- ^ A preprocessor if start line
   | PreIfEnd    String -- ^ A preprocessor if end line
```